collections/windowed.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Generates sliding views of the given array of the given size and returns a new
+ * array containing all of them.
+ *
+ * If step is set, each window will start that many elements after the last
+ * window's start. (Default: 1)
+ *
+ * If partial is set, windows will be generated for the last elements of the
+ * collection, resulting in some undefined values if size is greather than 1.
+ * (Default: false)
+ *
+ * Example:
+ *
+ * ```ts
+ * import { windowed } from "./windowed.ts"
+ * import { assertEquals } from "../testing/asserts.ts";
+ * const numbers = [ 1, 2, 3, 4, 5 ]
+ *
+ * const windows = windowed(numbers, 3)
+ * assertEquals(windows, [
+ *     [ 1, 2, 3 ],
+ *     [ 2, 3, 4 ],
+ *     [ 3, 4, 5 ],
+ * ])
+ *
+ * const windowsWithStep = windowed(numbers, 3, {step: 2})
+ * assertEquals(windowsWithStep, [
+ *     [ 1, 2, 3 ],
+ *     [ 3, 4, 5 ],
+ * ])
+ *
+ * const windowsWithPartial = windowed(numbers, 3, {partial: true})
+ * assertEquals(windowsWithPartial, [
+ *     [ 1, 2, 3 ],
+ *     [ 2, 3, 4 ],
+ *     [ 3, 4, 5 ],
+ *     [4, 5],
+ *     [5],
+ * ])
+ * ```
+ */
+export function windowed<T>(
+  array: readonly T[],
+  size: number,
+  { step = 1, partial = false }: {
+    /**
+     * If step is set, each window will start that many elements after the last
+     * window's start. (Default: 1)
+     */
+    step?: number;
+    /**
+     * If partial is set, windows will be generated for the last elements of the
+     * collection, resulting in some undefined values if size is greather than 1.
+     * (Default: false)
+     */
+    partial?: boolean;
+  } = {},
+): T[][] {
+  if (
+    !Number.isInteger(size) || !Number.isInteger(step) || size <= 0 || step <= 0
+  ) {
+    throw new RangeError("Both size and step must be positive integer.");
+  }
+
+  /** length of the return array */
+  const length = Math.floor((array.length - (partial ? 1 : size)) / step + 1);
+
+  const result = [];
```

Comment:
```suggestion
  const result = new Array(length);
```

Reply:
Proposed Thank you.
Sure, `array.push` (0.36173926999999984ms) is a bit faster than` new Array(length)` (0.33261830000000053ms).
However, according to [this V8 blog post](https://v8.dev/blog/elements-kinds), it seems that V8 can optimize the code better with `array.push` than with` new Array(length)`. I would like to use `array.push` to improve the performance of subsequent code that uses the return value of this function.

Reply:
you are pushing arrays into array, primitive value optimization is impossible in this case

Reply:
@devcat what @ayame113 is (I think correctly) getting at, is that when you build an Array using the constructor with a length value, you are saving some performance for reallocation when filling it, but that Array is and will forever be markes as `HOLEY_`, which prevents v8 from applying performance optimizations to further operations on it like `.map()`.

I think this is actually really interesting here **and** in all the other functions. We have been doing `new Array(length)` whenever possible, but thinking about it now, that means that we improve the performance of our function a bit, but `return` an unoptimizable `HOLEY_` Array to the user that will perform worse in their code. Specifically for bigger Arrays, I don't think that is desirable, as the reallocation impact is only `O(log n)` and only happens once, while the impact of the user carrying a `HOLEY_` Array through their code happens ok every single further operation on it.

I plan to do a benchmark suite for the whole module as soons as we are API complete, we should investigate that further once that lands.

Reply:
@LionC 
Thanks for the explanation, that's what I wanted to say. 👍